gnmidiff/setrequest.go

```diff
@@ -267,52 +271,173 @@ func minimalSetRequestIntent(req *gpb.SetRequest) (setRequestIntent, error) {
 	return intent, nil
 }
 
+// binaryBase64 takes an input byte slice and returns it as a base64
+// encoded string.
+func binaryBase64(i []byte) string {
+	var b bytes.Buffer
+	encoder := base64.NewEncoder(base64.StdEncoding, &b)
+	encoder.Write(i)
+	encoder.Close()
+	return b.String()
+}
+
+// protoLeafToJSON converts a gNMI proto scalar to the equivalent RFC7951 JSON
+// representation.
+func protoLeafToJSON(tv *gpb.TypedValue) (interface{}, error) {
+	switch tv.GetValue().(type) {
+	case *gpb.TypedValue_StringVal:
+		return tv.GetStringVal(), nil
+	case *gpb.TypedValue_BoolVal:
+		return tv.GetBoolVal(), nil
+	case *gpb.TypedValue_IntVal:
+		return float64(tv.GetIntVal()), nil
+	case *gpb.TypedValue_UintVal:
+		return float64(tv.GetUintVal()), nil
+	case *gpb.TypedValue_DoubleVal:
+		return float64(tv.GetDoubleVal()), nil
+	case *gpb.TypedValue_LeaflistVal:
+		elems := tv.GetLeaflistVal().GetElement()
+		ss := make([]interface{}, len(elems))
+		for x, e := range elems {
+			var err error
+			if ss[x], err = protoLeafToJSON(e); err != nil {
+				return nil, err
+			}
+		}
+		return ss, nil
+	case *gpb.TypedValue_BytesVal:
+		return binaryBase64(tv.GetBytesVal()), nil
+	default:
+		return nil, fmt.Errorf("gnmidiff: TypedValue type %T is not a scalar type", tv.GetValue())
+	}
+}
+
 // populateUpdates populates all leaf updates at the given path into the intent.
 //
 // Note: The input path must NOT end with "/".
 //
 // e.g. /a for b/c="foo" would introduce an update of /a/b/c="foo" into the intent.
-func populateUpdates(intent *setRequestIntent, path string, tv *gpb.TypedValue) error {
+func populateUpdates(intent *setRequestIntent, path string, tv *gpb.TypedValue, newSchemaFn func() (*ytypes.Schema, error)) error {
 	if len(path) > 0 && path[len(path)-1] == '/' {
 		return fmt.Errorf("gnmidiff: invalid input path %q, must not end with \"/\"", path)
 	}
-	var leafVal interface{}
-	isLeaf := true
-	switch tv.GetValue().(type) {
-	// TODO: Handle other scalar types.
-	case *gpb.TypedValue_StringVal:
-		leafVal = tv.GetStringVal()
-	case *gpb.TypedValue_BoolVal:
-		leafVal = tv.GetBoolVal()
-	case *gpb.TypedValue_JsonIetfVal:
-		updates, err := flattenOCJSON(tv.GetJsonIetfVal(), false)
-		if err != nil {
-			return err
-		}
-		if val, ok := updates[""]; len(updates) == 1 && ok {
-			leafVal = val
-		} else {
-			isLeaf = false
-			for subpath, value := range updates {
-				pathToLeaf := path + subpath
-				if _, ok := intent.Updates[pathToLeaf]; ok {
-					return fmt.Errorf("gnmidiff: leaf value set twice in SetRequest: %v", pathToLeaf)
+
+	// Populate updates when the schema is unknown.
+	if newSchemaFn == nil {
```

Comment:
Would it be possible to put lines 327-365 into a function (say unknownSchemaUpdate) and then here just say
```
if newSchema == nil {
    return unknownSchemaUpdate(...)
}
```

I think that will make this function. more readable.

Reply:
Done. Also added a comment block to explain how the schema-aware updates are flattened.